File: src/2025-02-27-deepclaude37-rust/cpp/exercises/practice/grade-school/grade_school.cpp
#include "grade_school.h"

namespace grade_school {

void school::add(const std::string& name, int grade) {
    auto& students = students_by_grade_[grade];
    students.insert(name);
}

std::vector<std::string> school::grade(int grade) const {
    auto it = students_by_grade_.find(grade);
    if (it != students_by_grade_.end()) {
        return {it->second.begin(), it->second.end()};
    }
    return {};
}

std::vector<std::string> school::roster() const {
    std::vector<std::string> all_students;
    for (const auto& [grade_num, students] : students_by_grade_) {
        all_students.insert(all_students.end(), students.begin(), students.end());
    }
    return all_students;
}

}  // namespace grade_school

File: src/python-2025-02-24-deepclaude-rust/python/exercises/practice/grade-school/grade_school.py
class School:
    def __init__(self):
        self._student_grades = {}
        self._grade_students = {}
        self._added_list = []

    def add_student(self, name, grade):
        if name in self._student_grades:
            self._added_list.append(False)
            return
        self._student_grades[name] = grade
        if grade not in self._grade_students:
            self._grade_students[grade] = set()
        self._grade_students[grade].add(name)
        self._added_list.append(True)

    def roster(self):
        sorted_grades = sorted(self._grade_students.keys())
        result = []
        for grade in sorted_grades:
            students = sorted(self._grade_students[grade])
            result.extend(students)
        return result

    def grade(self, grade_number):
        students = self._grade_students.get(grade_number, set())
        return sorted(students)

    def added(self):
        return self._added_list
